File: ios/App/App/RoadSnap.swift
import Foundation
import CoreLocation

/// SNAP-TO-PATH nativo iOS — port fedele di src/lib/roadSnap.ts e RoadSnap.kt.
///
/// Scarica la geometria strade/marciapiedi di un'area da /api/roads/tile, la
/// indicizza a griglia in RAM e "snappa" la posizione GPS sul segmento
/// percorribile piu' vicino in modo CONSERVATIVO: solo se una strada e' entro
/// la soglia, altrimenti ritorna nil e il chiamante usa il GPS grezzo. Al
/// momento dello snap non serve rete (indice gia' in memoria).
///
/// NB: NON testato sul campo. Conservativo per costruzione (worst case = GPS
/// grezzo). Già nel target Xcode "App" (Sources build phase in
/// project.pbxproj, come TriggerTelemetry.swift) — questo commento diceva
/// il contrario ma era rimasto stantio da quando il file non c'era ancora.
final class RoadSnap {
    static let shared = RoadSnap()
    private init() { loadCached() }

    // Persistenza: il tile dell'area visitata online resta disponibile offline
    // dopo un riavvio. Un tile di un'altra area è innocuo (celle non combaciano).
    private var cacheURL: URL? {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first?
            .appendingPathComponent("road_tile.json")
    }

    private let cell = 0.003 // ~300 m
    private let roadsUrl = "https://wip.guide/api/roads/tile"

    private struct Seg { let aLat: Double; let aLon: Double; let bLat: Double; let bLon: Double }

    /// Cella della griglia. Era una `String` interpolata ("12345,678"): a ogni
    /// segmento indicizzato e a ogni lookup si costruiva e si hashava una
    /// stringa, cioè un'allocazione — nove per ogni snap, decine di migliaia
    /// alla costruzione dell'indice. Due Int in una struct Hashable fanno lo
    /// stesso lavoro senza toccare l'heap.
    private struct Cell: Hashable { let x: Int; let y: Int }

    private var carGrid: [Cell: [Seg]] = [:]
    private var footGrid: [Cell: [Seg]] = [:]
    private var haveTile = false
    private var lastLat = 0.0
    private var lastLon = 0.0
    // Guardia di concorrenza: una sola richiesta in volo (protetta da lock).
    private var fetching = false
    private let lock = NSLock()

    // Esito dell'ULTIMO tentativo, riuscito o fallito (prima si registrava solo
    // il successo: con la rete giu' si riprovava a ogni fix GPS).
    private var lastAttemptAt: TimeInterval = 0   // ProcessInfo.systemUptime, monotono
    private var failStreak = 0                    // fallimenti consecutivi
    // Modo di trasporto dell'ultimo snap(): il chiamante non lo passa a
    // shouldRefresh(), quindi lo memorizziamo qui. Default "auto" = soglia piu'
    // stretta, il caso conservativo.
    private var lastDriving = true

    // Attesa crescente dopo un fallimento (5 s, 15 s, 60 s, 5 min, tetto 15 min).
    private let backoff: [TimeInterval] = [5, 15, 60, 300, 900]
    // Intervallo minimo fra due download riusciti: guardia contro fix GPS che
    // saltano avanti e indietro (a 25 m/s la soglia auto si copre in ~18 s).
    private let minInterval: TimeInterval = 20

    // Soglie di rinfresco per modo di trasporto. Il tile ha raggio 700 m
    // centrato sul punto in cui e' stato scaricato: dopo uno spostamento di d
    // il bordo davanti a noi e' a 700 - d metri, ed e' quel margine che deve
    // bastare a completare il download successivo.
    //  - A piedi (1,4 m/s): soglia 550 m -> margine 150 m = ~107 s di cammino,
    //    e un download ogni 550/1,4 = ~393 s, cioe' ~9 all'ora (prima, con 400 m,
    //    erano ~13/ora). Il raggio 700 m copre comodamente i tempi pedonali.
    //  - In auto (14 m/s in citta'/extraurbano): soglia 450 m -> margine 250 m
    //    = ~18 s a 14 m/s e ~10 s a 25 m/s in autostrada, quindi il download
    //    parte prima di uscire dalla tile; un download ogni 450/14 = ~32 s,
    //    cioe' ~112 all'ora contro i ~126 dei 400 m fissi. Se il margine non
    //    basta lo snap semplicemente non si applica (fail-open, GPS grezzo).
    private let thresholdFootM = 550.0
    private let thresholdCarM = 450.0

    private func cellKey(_ lat: Double, _ lon: Double) -> Cell {
        return Cell(x: Int((lat / cell).rounded()), y: Int((lon / cell).rounded()))
    }

    private func metersBetween(_ lat1: Double, _ lon1: Double, _ lat2: Double, _ lon2: Double) -> Double {
        let r = 6371000.0
        let dLat = (lat2 - lat1) * .pi / 180
        let dLon = (lon2 - lon1) * .pi / 180
        let a = sin(dLat / 2) * sin(dLat / 2) +
            cos(lat1 * .pi / 180) * cos(lat2 * .pi / 180) * sin(dLon / 2) * sin(dLon / 2)
        return 2 * r * asin(min(1.0, sqrt(a)))
    }

    private func projectToSeg(_ lat: Double, _ lon: Double, _ s: Seg) -> (lat: Double, lon: Double, distM: Double) {
        var cosLat = cos(lat * .pi / 180); if cosLat == 0 { cosLat = 1 }
        let px = lon * cosLat, py = lat
        let ax = s.aLon * cosLat, ay = s.aLat, bx = s.bLon * cosLat, by = s.bLat
        let dx = bx - ax, dy = by - ay
        let len2 = dx * dx + dy * dy
        var t = len2 == 0 ? 0 : ((px - ax) * dx + (py - ay) * dy) / len2
        t = max(0, min(1, t))
        let snapLat = ay + t * dy
        let snapLon = (ax + t * dx) / cosLat
        return (snapLat, snapLon, metersBetween(lat, lon, snapLat, snapLon))
    }

    /// Snap conservativo: ritorna una CLLocation snappata o nil (usa GPS grezzo).
    func snap(_ loc: CLLocation, isDriving: Bool) -> CLLocation? {
        lock.lock()
        lastDriving = isDriving // memorizzato per la soglia di shouldRefresh()
        let grid = isDriving ? carGrid : footGrid
        lock.unlock()
        if grid.isEmpty { return nil }
        let lat = loc.coordinate.latitude, lon = loc.coordinate.longitude
        let acc = loc.horizontalAccuracy
        let maxSnap = min(40.0, max(20.0, acc <= 0 ? 20.0 : acc))
        let cLat = Int((lat / cell).rounded()), cLon = Int((lon / cell).rounded())
        var best: (lat: Double, lon: Double, distM: Double)?
        for dLa in -1...1 {
            for dLo in -1...1 {
                guard let arr = grid[Cell(x: cLat + dLa, y: cLon + dLo)] else { continue }
                for s in arr {
                    let p = projectToSeg(lat, lon, s)
                    if best == nil || p.distM < best!.distM { best = p }
                }
            }
        }
        guard let b = best, b.distM <= maxSnap else { return nil }
        if metersBetween(lat, lon, b.lat, b.lon) < 3.0 { return nil }
        return CLLocation(
            coordinate: CLLocationCoordinate2D(latitude: b.lat, longitude: b.lon),
            altitude: loc.altitude, horizontalAccuracy: loc.horizontalAccuracy,
            verticalAccuracy: loc.verticalAccuracy, course: loc.course,
            speed: loc.speed, timestamp: loc.timestamp
        )
    }

    /// Serve un nuovo tile? thresholdM <= 0 (il default, quello che usa il
    /// manager) sceglie la soglia dal modo di trasporto dell'ultimo snap().
    /// Durante l'attesa dopo un fallimento ritorna false senza scrivere log: lo
    /// snap semplicemente non si applica e si usa il fix grezzo (fail-open).
    func shouldRefresh(_ loc: CLLocation, thresholdM: Double = 0) -> Bool {
        lock.lock()
        let inFlight = fetching
        let have = haveTile, lLat = lastLat, lLon = lastLon
        let attemptAt = lastAttemptAt, streak = failStreak, driving = lastDriving
        lock.unlock()
        if inFlight { return false } // richiesta gia' in volo
        if attemptAt > 0 {
            let wait = streak > 0 ? backoff[min(streak - 1, backoff.count - 1)] : minInterval
            if ProcessInfo.processInfo.systemUptime - attemptAt < wait { return false }
        }
        if !have { return true }
        let th = thresholdM > 0 ? thresholdM : (driving ? thresholdCarM : thresholdFootM)
        return metersBetween(loc.coordinate.latitude, loc.coordinate.longitude, lLat, lLon) > th
    }

    /// Registra SEMPRE l'esito del tentativo (riuscito o fallito) e libera la
    /// guardia di concorrenza. Il contatore dei fallimenti pilota l'attesa
    /// crescente e si azzera al primo successo.
    private func finishAttempt(ok: Bool) {
        lock.lock()
        lastAttemptAt = ProcessInfo.processInfo.systemUptime
        let first = failStreak == 0
        failStreak = ok ? 0 : min(failStreak + 1, backoff.count)
        fetching = false
        lock.unlock()
        // Un solo log per serie di fallimenti: niente log a raffica.
        if !ok && first { NSLog("[RoadSnap] refresh tile strade fallito") }
    }

    /// Scarica e reindicizza il tile (async, best-effort).
    func refresh(_ loc: CLLocation, radius: Int = 700) {
        lock.lock(); if fetching { lock.unlock(); return }; fetching = true; lock.unlock()
        let lat = loc.coordinate.latitude, lon = loc.coordinate.longitude
        guard let url = URL(string: "\(roadsUrl)?lat=\(lat)&lon=\(lon)&radius=\(radius)") else {
            finishAttempt(ok: false); return
        }
        URLSession.shared.dataTask(with: url) { [weak self] data, response, _ in
            guard let self = self else { return }
            let httpOk = (response as? HTTPURLResponse).map { (200...299).contains($0.statusCode) } ?? true
            guard httpOk, let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                self.finishAttempt(ok: false); return
            }
            let car = self.buildGrid(json["car"] as? [[[Double]]])
            let foot = self.buildGrid(json["foot"] as? [[[Double]]])
            // Un tile vuoto (risposta malformata o senza strade) non deve buttare
            // via l'indice buono che abbiamo gia' in RAM: meglio tenerlo e
            // trattare il tentativo come fallito.
            if car.isEmpty && foot.isEmpty { self.finishAttempt(ok: false); return }
            self.lock.lock()
            self.carGrid = car; self.footGrid = foot
            self.lastLat = lat; self.lastLon = lon; self.haveTile = true
            self.lock.unlock()
            if let u = self.cacheURL { try? data.write(to: u) } // persisti per l'offline
            self.finishAttempt(ok: true)
        }.resume()
    }

    /// Carica il tile persistito (offline). Chiamato all'init.
    private func loadCached() {
        guard let u = cacheURL, let data = try? Data(contentsOf: u),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }
        let car = buildGrid(json["car"] as? [[[Double]]])
        let foot = buildGrid(json["foot"] as? [[[Double]]])
        lock.lock(); carGrid = car; footGrid = foot; haveTile = true; lock.unlock()
    }

    private func buildGrid(_ polys: [[[Double]]]?) -> [Cell: [Seg]] {
        var g: [Cell: [Seg]] = [:]
        guard let polys = polys else { return g }
        for poly in polys {
            var j = 0
            while j + 1 < poly.count {
                let p0 = poly[j], p1 = poly[j + 1]
                if p0.count >= 2 && p1.count >= 2 {
                    let seg = Seg(aLat: p0[0], aLon: p0[1], bLat: p1[0], bLon: p1[1])
                    // Il segmento va indicizzato SOTTO LE CELLE CHE TOCCA, e
                    // quasi sempre è una sola: la cella è ~300 m, i segmenti
                    // stradali di OSM sono decine di metri. Prima si faceva
                    // comunque un doppio inserimento (entrambi gli estremi),
                    // quindi la stessa lista conteneva due volte lo stesso
                    // segmento e lo snap lo proiettava due volte. La seconda
                    // cella si aggiunge solo se è davvero diversa: la
                    // copertura resta identica, il lavoro si dimezza.
                    let cellaA = cellKey(seg.aLat, seg.aLon)
                    g[cellaA, default: []].append(seg)
                    let cellaB = cellKey(seg.bLat, seg.bLon)
                    if cellaB != cellaA { g[cellaB, default: []].append(seg) }
                }
                j += 1
            }
        }
        return g
    }
}

File: android/app/src/main/java/com/itaintasca/app/geofence/RoadSnap.kt
package com.itaintasca.app.geofence

import android.os.SystemClock
import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * SNAP-TO-PATH nativo — port fedele di src/lib/roadSnap.ts.
 *
 * Scarica la geometria strade/marciapiedi di un'area da /api/roads/tile, la
 * indicizza a griglia in RAM e "snappa" la posizione GPS sul segmento
 * percorribile piu' vicino in modo CONSERVATIVO: solo se una strada e' entro
 * la soglia, altrimenti ritorna null e il chiamante usa il GPS grezzo. Al
 * momento dello snap non serve rete (l'indice e' gia' in memoria).
 *
 * NB: NON testato sul campo. Conservativo per costruzione (worst case =
 * comportamento attuale). Il fetch va fatto fuori dal main thread.
 */
object RoadSnap {
    private const val TAG = "RoadSnap"
    private const val CELL = 0.003 // ~300 m
    // (SEC-09) Dominio unico in WipApi.BASE.
    private const val ROADS_URL = com.itaintasca.app.service.WipApi.BASE + "/api/roads/tile"

    private data class Seg(val aLat: Double, val aLon: Double, val bLat: Double, val bLon: Double)

    @Volatile private var carGrid: Map<String, List<Seg>> = emptyMap()
    @Volatile private var footGrid: Map<String, List<Seg>> = emptyMap()
    @Volatile private var haveTile = false
    @Volatile private var lastLat = 0.0
    @Volatile private var lastLon = 0.0

    // Guardia di concorrenza: una sola richiesta in volo. AtomicBoolean e non
    // un @Volatile Boolean perche' il "leggi-e-poi-scrivi" fra due thread IO
    // lascerebbe passare due download simultanei.
    private val fetching = AtomicBoolean(false)

    // Esito dell'ULTIMO tentativo, riuscito o fallito (prima si registrava solo
    // il successo: con la rete giu' si riprovava a ogni fix GPS).
    @Volatile private var lastAttemptAt = 0L      // SystemClock.elapsedRealtime()
    @Volatile private var failStreak = 0          // fallimenti consecutivi
    // Modo di trasporto dell'ultimo snap(): il chiamante non lo passa a
    // shouldRefresh(), quindi lo memorizziamo qui. Default "auto" = soglia piu'
    // stretta, il caso conservativo.
    @Volatile private var lastCar = true

    // Attesa crescente dopo un fallimento (5 s, 15 s, 60 s, 5 min, tetto 15 min).
    private val BACKOFF_MS = longArrayOf(5_000L, 15_000L, 60_000L, 300_000L, 900_000L)
    // Intervallo minimo fra due download riusciti: guardia contro fix GPS che
    // saltano avanti e indietro (a 25 m/s la soglia auto si copre in ~18 s).
    private const val MIN_INTERVAL_MS = 20_000L

    // Soglie di rinfresco per modo di trasporto. Il tile ha raggio 700 m
    // centrato sul punto in cui e' stato scaricato: dopo uno spostamento di d
    // il bordo davanti a noi e' a 700 - d metri, ed e' quel margine che deve
    // bastare a completare il download successivo.
    //  - A piedi (1,4 m/s): soglia 550 m -> margine 150 m = ~107 s di cammino,
    //    e un download ogni 550/1,4 = ~393 s, cioe' ~9 all'ora (prima, con 400 m,
    //    erano ~13/ora). Il raggio 700 m copre comodamente i tempi pedonali.
    //  - In auto (14 m/s in citta'/extraurbano): soglia 450 m -> margine 250 m
    //    = ~18 s a 14 m/s e ~10 s a 25 m/s in autostrada, quindi il download
    //    parte prima di uscire dalla tile; un download ogni 450/14 = ~32 s,
    //    cioe' ~112 all'ora contro i ~126 dei 400 m fissi. Se il margine non
    //    basta lo snap semplicemente non si applica (fail-open, GPS grezzo).
    private const val THRESHOLD_FOOT_M = 550.0
    private const val THRESHOLD_CAR_M = 450.0

    // Persistenza: il tile dell'area visitata online resta disponibile offline
    // dopo un riavvio. Un tile di un'altra area è innocuo (celle non combaciano
    // → nessuno snap, GPS grezzo). Impostare cacheDir a filesDir all'avvio.
    @Volatile var cacheDir: java.io.File? = null
    private const val CACHE_FILE = "road_tile.json"

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private fun cellKey(lat: Double, lon: Double) = "${Math.round(lat / CELL)},${Math.round(lon / CELL)}"

    private fun metersBetween(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val r = 6371000.0
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val a = sin(dLat / 2).pow(2) +
            cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) * sin(dLon / 2).pow(2)
        return 2 * r * asin(min(1.0, sqrt(a)))
    }

    private class Proj(val lat: Double, val lon: Double, val distM: Double)

    private fun projectToSeg(lat: Double, lon: Double, s: Seg): Proj {
        val cosLat = cos(Math.toRadians(lat)).let { if (it == 0.0) 1.0 else it }
        val px = lon * cosLat; val py = lat
        val ax = s.aLon * cosLat; val ay = s.aLat
        val bx = s.bLon * cosLat; val by = s.bLat
        val dx = bx - ax; val dy = by - ay
        val len2 = dx * dx + dy * dy
        var t = if (len2 == 0.0) 0.0 else ((px - ax) * dx + (py - ay) * dy) / len2
        t = max(0.0, min(1.0, t))
        val snapLat = ay + t * dy
        val snapLon = (ax + t * dx) / cosLat
        return Proj(snapLat, snapLon, metersBetween(lat, lon, snapLat, snapLon))
    }

    /** Snap conservativo. Ritorna (lat,lon) snappati oppure null (usa GPS grezzo). */
    fun snap(lat: Double, lon: Double, accM: Float, isCar: Boolean): Pair<Double, Double>? {
        lastCar = isCar // memorizzato per la soglia di shouldRefresh()
        val grid = if (isCar) carGrid else footGrid
        if (grid.isEmpty()) return null
        val maxSnap = min(40.0, max(20.0, if (accM <= 0f) 20.0 else accM.toDouble()))
        val cLat = Math.round(lat / CELL)
        val cLon = Math.round(lon / CELL)
        var best: Proj? = null
        for (dLa in -1..1) for (dLo in -1..1) {
            val arr = grid["${cLat + dLa},${cLon + dLo}"] ?: continue
            for (s in arr) {
                val p = projectToSeg(lat, lon, s)
                if (best == null || p.distM < best!!.distM) best = p
            }
        }
        val b = best ?: return null
        if (b.distM > maxSnap) return null
        if (metersBetween(lat, lon, b.lat, b.lon) < 3.0) return null
        return Pair(b.lat, b.lon)
    }

    /**
     * Serve un nuovo tile? thresholdM <= 0 (il default, quello che usa il
     * servizio) sceglie la soglia dal modo di trasporto dell'ultimo snap().
     * Durante l'attesa dopo un fallimento ritorna false senza scrivere log: lo
     * snap semplicemente non si applica e si usa il fix grezzo (fail-open).
     */
    fun shouldRefresh(lat: Double, lon: Double, thresholdM: Double = 0.0): Boolean {
        if (fetching.get()) return false // richiesta gia' in volo
        val now = SystemClock.elapsedRealtime()
        if (lastAttemptAt != 0L) {
            val wait = if (failStreak > 0)
                BACKOFF_MS[min(failStreak - 1, BACKOFF_MS.size - 1)]
            else MIN_INTERVAL_MS
            if (now - lastAttemptAt < wait) return false
        }
        if (!haveTile) return true
        val th = if (thresholdM > 0.0) thresholdM
        else if (lastCar) THRESHOLD_CAR_M else THRESHOLD_FOOT_M
        return metersBetween(lat, lon, lastLat, lastLon) > th
    }

    /** Scarica e reindicizza il tile. Best-effort, BLOCCANTE (chiamare su IO). */
    fun refresh(lat: Double, lon: Double, radius: Int = 700) {
        if (!fetching.compareAndSet(false, true)) return // una sola in volo
        var ok = false
        try {
            val url = "$ROADS_URL?lat=$lat&lon=$lon&radius=$radius"
            client.newCall(Request.Builder().url(url).build()).execute().use { resp ->
                val body = if (resp.isSuccessful) resp.body?.string() else null
                if (body != null) {
                    val json = JSONObject(body)
                    val car = buildGrid(json.optJSONArray("car"))
                    val foot = buildGrid(json.optJSONArray("foot"))
                    // Un tile vuoto (risposta malformata o senza strade) non deve
                    // buttare via l'indice buono che abbiamo gia' in RAM: meglio
                    // tenerlo e trattare il tentativo come fallito.
                    if (car.isNotEmpty() || foot.isNotEmpty()) {
                        carGrid = car; footGrid = foot
                        lastLat = lat; lastLon = lon; haveTile = true
                        cacheDir?.let { try { java.io.File(it, CACHE_FILE).writeText(body) } catch (_: Exception) {} }
                        Log.d(TAG, "Tile strade caricato: car=${car.size} celle, foot=${foot.size} celle")
                        ok = true
                    }
                }
            }
        } catch (e: Exception) {
            // Un solo log per serie di fallimenti: niente log a raffica.
            if (failStreak == 0) Log.w(TAG, "refresh fallito: ${e.message}")
        } finally {
            // L'esito si registra SEMPRE, non solo in caso di successo.
            lastAttemptAt = SystemClock.elapsedRealtime()
            failStreak = if (ok) 0 else min(failStreak + 1, BACKOFF_MS.size)
            fetching.set(false)
        }
    }

    /** Carica il tile persistito (offline). Da chiamare all'avvio dopo cacheDir. */
    fun loadCached() {
        val dir = cacheDir ?: return
        try {
            val f = java.io.File(dir, CACHE_FILE)
            if (!f.exists()) return
            val json = JSONObject(f.readText())
            carGrid = buildGrid(json.optJSONArray("car"))
            footGrid = buildGrid(json.optJSONArray("foot"))
            haveTile = true // shouldRefresh riscarica appena online ci si sposta
            Log.d(TAG, "Tile strade ripristinato da disco")
        } catch (_: Exception) { /* nessun tile persistito */ }
    }

    private fun buildGrid(polys: JSONArray?): Map<String, List<Seg>> {
        val g = HashMap<String, MutableList<Seg>>()
        if (polys == null) return g
        for (i in 0 until polys.length()) {
            val poly = polys.optJSONArray(i) ?: continue
            var j = 0
            while (j + 1 < poly.length()) {
                val p0 = poly.optJSONArray(j)
                val p1 = poly.optJSONArray(j + 1)
                if (p0 != null && p1 != null) {
                    val seg = Seg(p0.optDouble(0), p0.optDouble(1), p1.optDouble(0), p1.optDouble(1))
                    g.getOrPut(cellKey(seg.aLat, seg.aLon)) { mutableListOf() }.add(seg)
                    g.getOrPut(cellKey(seg.bLat, seg.bLon)) { mutableListOf() }.add(seg)
                }
                j++
            }
        }
        return g
    }
}
